didcomm/plaintext.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Optional, List, Union, Dict
+
+from didcomm.common.types import JSON_DATA, DID, JSON, JWT, DID_URL
+from didcomm.did_doc.did_resolver import DIDResolver
+from didcomm.secrets.secrets_resolver import SecretsResolver
+
+Header = Dict[str, Union[str, int, JSON_DATA]]
+
+
+@dataclass
+class PlaintextOptionalHeaders:
+    """Optional headers for any Plaintext message"""
+    typ: Optional[str] = None
+    frm: Optional[DID] = None
+    to: Optional[List[DID]] = None
+    created_time: Optional[int] = None
+    expires_time: Optional[int] = None
+    from_prior: Optional[JWT] = None
+    please_ack: Optional[bool] = None
+    ack: Optional[List[str]] = None
+    thid: Optional[str] = None
+    pthid: Optional[str] = None
+    attachments: Optional[List[Attachment]] = None
+    custom_headers: Optional[List[Header]] = None
+
+
+@dataclass
+class PlaintextRequiredHeaders:
+    """Required headers for any Plaintext message"""
+    id: str
+    type: str
+
+
+@dataclass
+class PlaintextBody:
+    """Plaintext body as a application/protocol specific data"""
+    body: JSON_DATA
+
+
+@dataclass
+class Plaintext(PlaintextOptionalHeaders, PlaintextRequiredHeaders, PlaintextBody):
+    """Plaintext message consisting of headers and application/protocol specific data (body)"""
+
+    def to_json(self) -> JSON:
+        return ""
+
+
+@dataclass(frozen=True)
+class Attachment:
+    """Plaintext attachment"""
+    id: str
+    data: Union[AttachmentDataLinks, AttachmentDataBase64, AttachmentDataJson]
+    description: Optional[str] = None
+    filename: Optional[str] = None
+    mime_type: Optional[str] = None
+    format: Optional[str] = None
+    lastmod_time: Optional[int] = None
+    byte_count: Optional[int] = None
+
+
+@dataclass(frozen=True)
+class AttachmentDataLinks:
+    links: List[str]
+    hash: str
```

Comment:
I had a quick look in the spec and couldn't see anything obvious, but is there a reason that `hash` is mandatory here and not in the other `Attachment`s?

Reply:
Yes, it's explicitly required in the spec: https://identity.foundation/didcomm-messaging/spec/#reference-2
> The hash of the content encoded in multi-hash format. Used as an integrity check for the attachment, and MUST be used if the data is referenced via the links data attribute.